block_writer/main.go

```diff
@@ -179,38 +188,60 @@ func main() {
 	writers := make([]blockWriter, *concurrency)
 
 	errCh := make(chan error)
+	var wg sync.WaitGroup
 	for i := range writers {
+		wg.Add(1)
 		writers[i] = newBlockWriter(db)
-		go writers[i].run(errCh)
+		go writers[i].run(errCh, &wg)
 	}
 
 	var numErr int
-	for range time.Tick(*outputInterval) {
-		now := time.Now()
-		elapsed := time.Since(lastNow)
-		dumps := atomic.LoadUint64(&numBlocks)
-		fmt.Printf("%6s: %6.1f/sec",
-			time.Duration(time.Since(start).Seconds()+0.5)*time.Second,
-			float64(dumps-lastNumDumps)/elapsed.Seconds())
-		if numErr > 0 {
-			fmt.Printf(" (%d total errors)\n", numErr)
-		}
-		fmt.Printf("\n")
-		for {
-			select {
-			case err := <-errCh:
-				numErr++
-				if !*tolerateErrors {
-					log.Fatal(err)
-				} else {
-					log.Print(err)
-				}
-				continue
-			default:
+	tick := time.Tick(*outputInterval)
+	done := make(chan struct{}, 1)
+
+	go func() {
+		wg.Wait()
+		done <- struct{}{}
+	}()
+
+	if *duration > 0 {
+		go func() {
+			time.Sleep(*duration)
+			done <- struct{}{}
+		}()
+	}
+
+	for {
+		select {
+		case err := <-errCh:
+			numErr++
+			if !*tolerateErrors {
+				log.Fatal(err)
+			} else {
+				log.Print(err)
+			}
+			continue
+
+		case <-tick:
+			now := time.Now()
+			elapsed := time.Since(lastNow)
+			dumps := atomic.LoadUint64(&numBlocks)
+			fmt.Printf("%6s: %6.1f/sec",
+				time.Duration(time.Since(start).Seconds()+0.5)*time.Second,
+				float64(dumps-lastNumDumps)/elapsed.Seconds())
+			if numErr > 0 {
+				fmt.Printf(" (%d total errors)\n", numErr)
 			}
-			break
+			fmt.Printf("\n")
+			lastNumDumps = dumps
+			lastNow = now
+
+		case <-done:
+			dumps := atomic.LoadUint64(&numBlocks)
+			elapsed := time.Since(start).Seconds()
+			fmt.Printf("%6.1fs: %d total %6.1f/sec\n",
```

Comment:
Should probably log `numErr` here as well (in case we're running with `-tolerate-errors`)